### PR TITLE
Improve eBPF exit

### PR DESCRIPTION
**Pull request by thiagoftsm**
##### Summary

Sometimes eBPF.plugin was reporting these errors during shutdown:

```sh
#0  __GI___libc_free (mem=0x7f3a88224010) at malloc.c:3102
3102	malloc.c: No such file or directory.
[Current thread is 1 (Thread 0x7f3a888e6780 (LWP 131102))]
(gdb) bt
#0  __GI___libc_free (mem=0x7f3a88224010) at malloc.c:3102
#1  0x000055c8902c3cd1 in btf.free ()
#2  0x000055c89024e660 in ebpf_exit (sig=15) at collectors/ebpf.plugin/ebpf.c:473
#3  0x000055c89024ea75 in ebpf_stop_threads (sig=15) at collectors/ebpf.plugin/ebpf.c:553u use 
#4  <signal handler called>
#5  0x00007f3a889b4561 in __GI___clock_nanosleep (clock_id=0, flags=0, req=0x7ffc997622b0, rem=0x7ffc997622a0) at ../sysdeps/unix/sysv/linux/clock_nanosleep.c:48
#6  0x000055c89028ec42 in sleep_usec (usec=59999895) at libnetdata/clocks/clocks.c:350
#7  0x000055c89028e9c4 in heartbeat_next (hb=0x7ffc99762380, tick=60000000) at libnetdata/clocks/clocks.c:314
#8  0x000055c8902524d8 in main (argc=2, argv=0x7ffc997624a8) at collectors/ebpf.plugin/ebpf.c:2211
```

This PR is fixing SIGSEGV, SIGABRT, and double free corruption events found during more than 100 tests to shutdown plugin.

##### Test Plan
1. If you are not using `sytemd` (congratulations! :smile: ), set core dump size for `unlimited` (ulimit -S -c unlimited), but case you use `systemd` install `coredumpctl` on your environment.
2. Compile this branch
3. Start Netdata
4. Wait few minutes for all threads collect data.
5. Stop `netdata`
6. Verify threads stopped:
```sh
[root@archlinux netdata]# grep ebpf /var/log/netdata/error.log | grep finished
2022-11-23 04:11:01: ebpf.plugin INFO  : EBPF SOCKET : (0120@libnetdata/threads/t:thread_cleanup ): thread with task id 215524 finished
2022-11-23 04:11:01: ebpf.plugin INFO  : EBPF CACHESTAT : (0120@libnetdata/threads/t:thread_cleanup ): thread with task id 215525 finished
2022-11-23 04:11:01: ebpf.plugin INFO  : EBPF SYNC : (0120@libnetdata/threads/t:thread_cleanup ): thread with task id 215526 finished
2022-11-23 04:11:01: ebpf.plugin INFO  : EBPF DCSTAT : (0120@libnetdata/threads/t:thread_cleanup ): thread with task id 215527 finished
2022-11-23 04:11:01: ebpf.plugin INFO  : EBPF VFS : (0120@libnetdata/threads/t:thread_cleanup ): thread with task id 215529 finished
2022-11-23 04:11:01: ebpf.plugin INFO  : EBPF FILESYSTEM : (0120@libnetdata/threads/t:thread_cleanup ): thread with task id 215530 finished
2022-11-23 04:11:01: ebpf.plugin INFO  : EBPF MOUNT : (0120@libnetdata/threads/t:thread_cleanup ): thread with task id 215531 finished
2022-11-23 04:11:01: ebpf.plugin INFO  : EBPF FD : (0120@libnetdata/threads/t:thread_cleanup ): thread with task id 215532 finished
2022-11-23 04:11:01: ebpf.plugin INFO  : EBPF SOFTIRQ : (0120@libnetdata/threads/t:thread_cleanup ): thread with task id 215534 finished
2022-11-23 04:11:01: ebpf.plugin INFO  : EBPF OOMKILL : (0120@libnetdata/threads/t:thread_cleanup ): thread with task id 215535 finished
2022-11-23 04:11:01: ebpf.plugin INFO  : EBPF SHM : (0120@libnetdata/threads/t:thread_cleanup ): thread with task id 215536 finished
2022-11-23 04:11:01: ebpf.plugin INFO  : EBPF FS READ : (0120@libnetdata/threads/t:thread_cleanup ): thread with task id 215689 finished
2022-11-23 04:11:01: ebpf.plugin INFO  : EBPF HARDIRQ : (0120@libnetdata/threads/t:thread_cleanup ): thread with task id 215533 finished
2022-11-23 04:11:01: ebpf.plugin INFO  : EBPF SWAP : (0120@libnetdata/threads/t:thread_cleanup ): thread with task id 215528 finished
2022-11-23 04:11:01: ebpf.plugin INFO  : EBPF PROCESS : (0120@libnetdata/threads/t:thread_cleanup ): thread with task id 215523 finished
2022-11-23 04:11:01: ebpf.plugin INFO  : EBPF CGROUP : (0120@libnetdata/threads/t:thread_cleanup ): thread with task id 215552 finished
2022-11-23 04:11:01: ebpf.plugin INFO  : SYNC KERNEL : (0120@libnetdata/threads/t:thread_cleanup ): thread with task id 215645 finished
2022-11-23 04:11:01: ebpf.plugin INFO  : HARDIRQ KERNEL : (0120@libnetdata/threads/t:thread_cleanup ): thread with task id 215551 finished
2022-11-23 04:11:02: ebpf.plugin INFO  : SOFTIRQ KERNEL : (0120@libnetdata/threads/t:thread_cleanup ): thread with task id 215549 finished
2022-11-23 04:11:03: ebpf.plugin INFO  : DCSTAT KERNEL : (0120@libnetdata/threads/t:thread_cleanup ): thread with task id 215687 finished
2022-11-23 04:11:03: ebpf.plugin INFO  : CACHESTAT KERNEL : (0120@libnetdata/threads/t:thread_cleanup ): thread with task id 215634 finished
2022-11-23 04:11:04: ebpf.plugin INFO  : SWAP KERNEL : (0120@libnetdata/threads/t:thread_cleanup ): thread with task id 215619 finished
2022-11-23 04:11:04: ebpf.plugin INFO  : MOUNT KERNEL : (0120@libnetdata/threads/t:thread_cleanup ): thread with task id 215629 finished
2022-11-23 04:11:04: ebpf.plugin INFO  : SHM KERNEL : (0120@libnetdata/threads/t:thread_cleanup ): thread with task id 215648 finished
2022-11-23 04:11:06: ebpf.plugin INFO  : EBPF SOCKET READ : (0120@libnetdata/threads/t:thread_cleanup ): thread with task id 215688 finished
2022-11-23 04:11:06: ebpf.plugin INFO  : FD KERNEL : (0120@libnetdata/threads/t:thread_cleanup ): thread with task id 215635 finished
2022-11-23 04:11:06: ebpf.plugin INFO  : VFS KERNEL : (0120@libnetdata/threads/t:thread_cleanup ): thread with task id 215685 finished
```
7.  Verify that you do not have core dumps.
8. Now compile `master` branch and execute steps 3-6 again, you will see that you o not have the same number of threads reported.

##### Additional Information
It is not necessary to test on different kernels.

<details> <summary>For users: How does this change affect me?</summary>
Describe the PR affects users: 
- Which area of Netdata is affected by the change? ebpf.plugin
- Can they see the change or is it an under the hood? If they can see it, where? when stop ebpf.plugiin, plugin will close smothly.
- How is the user impacted by the change? a better plugin
- What are there any benefits of the change? Plugin Improvements
</details>

### CODE REVIEW

File: collectors/ebpf.plugin/ebpf.c [hunk 1]
```diff
@@ -27,8 +27,10 @@ struct config collector_config = { .first_section = NULL,
 int running_on_kernel = 0;
 int ebpf_nprocs;
 int isrh = 0;
+int main_thread_id = 0;
 
 pthread_mutex_t lock;
+pthread_mutex_t ebpf_exit_cleanup;
```
Comment (MrZammler) [inline collectors/ebpf.plugin/ebpf.c:33]: Hi Thiago!

Should there be a call for this for `pthread_mutex_init` ? (Or I couldn't see?)

File: collectors/ebpf.plugin/ebpf.c [hunk 2]
```diff
@@ -513,44 +516,68 @@ int ebpf_exit_plugin = 0;
...
+    if (main_thread_id != gettid() || only_one) {
+        pthread_mutex_unlock(&ebpf_exit_cleanup);
+        return;
+    }
+    only_one = 1;
+    pthread_mutex_unlock(&ebpf_exit_cleanup);
+
     int i;
-    for (i = 0; ebpf_threads[i].name != NULL; i++);
+    pthread_mutex_lock(&ebpf_exit_cleanup);
```
Comment (MrZammler) [inline collectors/ebpf.plugin/ebpf.c:531]: Any particular reason why it unlocks, then locks again after? Could it keep the lock up until 536?

File: collectors/ebpf.plugin/ebpf.c (hunk 1 above)
Comment (thiagoftsm) [reply collectors/ebpf.plugin/ebpf.c:33]: I was taking a look and I observed that I initialized it in a test branch, but I did not bring it here. Thanks! :man_facepalming:

File: collectors/ebpf.plugin/ebpf.c (hunk 2 above)
Comment (thiagoftsm) [reply collectors/ebpf.plugin/ebpf.c]: My bad, probably it was a commit during the dawn time. I moved the code, thanks!

File: collectors/ebpf.plugin/ebpf_socket.c [hunk 3]
```diff
@@ -2153,10 +2167,10 @@ void *ebpf_socket_read_hash(void *ptr)
...
     int fd_ipv6 = socket_maps[NETDATA_SOCKET_TABLE_IPV6].map_fd;
     int network_connection = em->optional;
-    while (ebpf_socket_exited == NETDATA_THREAD_EBPF_RUNNING) {
-        usec_t dt = heartbeat_next(&hb, step);
-        (void)dt;
-        if (ebpf_socket_exited == NETDATA_THREAD_EBPF_STOPPING)
+    while (!ebpf_exit_plugin) {
+        (void)heartbeat_next(&hb, step);
+
+        if (ebpf_exit_plugin)
```
Comment (MrZammler) [inline collectors/ebpf.plugin/ebpf_socket.c:2173]: Hey Thiago! Small one, maybe this if is not needed?

Comment (thiagoftsm) [reply collectors/ebpf.plugin/ebpf_socket.c]: I agree that was necessary with previous logic, I removed it with latest commit. Thanks!

**Review by MrZammler — APPROVED** [review]

**Review by underhood — APPROVED** [review]

**Pull request merged**